### ISSUE TITLE
Surfacing custom error message from server in build logs.

### ISSUE THREAD

**Issue by assimd** [gradle/gradle#18436]
### Expected Behavior

When a user tries to access a package from a repository, there could be multiple reason for that package to be not accessible. When such error occurs, we want to surface up actionable error messages that can reduce the user time in resolving the issue and overall improve the usability, like in the example below, we have added a message with link that user can follow to fix their issue.

```
* What went wrong:
Execution failed for task ':compileJava'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve foo:bar:10.+.
     Required by:
         project :
      > Failed to list versions for junit:junit.
         > Unable to load Maven meta-data from https://<redacted>/foo/bar/maven-metadata.xml.
            > Could not get resource 'https://<redacted>/foo/bar/maven-metadata.xml'.
               > Could not GET 'https://<redacted>/foo/bar/maven-metadata.xml'. Received status code 403 from server: Forbidden
               > With message: To request access to this package, click here: https://example.com/some/internal/portal?package=foo/bar

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
```
_Note: This is just a mock expectation, actual message does not have to be displayed as here_

### Current Behavior

```
* What went wrong:
Execution failed for task ':compileJava'.
> Could not resolve all files for configuration ':compileClasspath'.
   > Could not resolve foo:bar:10.+.
     Required by:
         project :
      > Failed to list versions for junit:junit.
         > Unable to load Maven meta-data from https://<redacted>/foo/bar/maven-metadata.xml.
            > Could not get resource 'https://<redacted>/foo/bar/maven-metadata.xml'.
               > Could not GET 'https://<redacted>/foo/bar/maven-metadata.xml'. Received status code 403 from server: Forbidden

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
* Get more help at https://help.gradle.org
```

As seen above, Gradle only display the HTTP status code, which is not enough to know the root case. A package may not be available due to various reasons, such as:
* Asset does not exist
* Package version does not exist
* Package does not exist
* Repository configured does not exist
* User does not have permission to access that package
* User does not have permission to access the repository
* User is not authenticated
* The package has been blocked for consumption
* Use may have to take some action to access the package (e.g. run some command or change some settings)

### Context

We have a custom maven repository solution for our internal developers to enable access to one or more [CodeArtifact repository](https://docs.aws.amazon.com/codeartifact/latest/ug/welcome.html) for their builds. The system integrates with our internal authentication mechanism and there can be various reasons when a developer is not able to access the package or a repository (some of them listed above) and as the error message does not have sufficient information, it generally results in user confusion and them spending time in search for the solution. To help our developers, we want to surface a custom error message supplied by the server as part of build error message, so users can quickly find the relevant information.

We explored options of using [`UnresolvedDependencyResult`](https://docs.gradle.org/current/javadoc/org/gradle/api/artifacts/result/UnresolvedDependencyResult.html), but it does not contain any data related to response from the server, so it is not usable for our use case.

We understand that this feature request very specific to our use case, but if implemented in a generic way, other repository solutions like Sonatype and Nexus can also benefit from this feature. If there is already a design in place for such feature, we can also contribute towards its implementation. Let us know your thoughts on this.

**Comment by vlsi** [comment]
Just in case, if authentication is required, then non-authenticated users must not be able to tell the difference between `Package does not exist` and `User is not authenticated`.
In other words, if a non-authenticated user is able to tell which packages (or their versions) exist in the repository, then it is a security issue.

**Comment by assimd** [comment]
@vlsi - I agree, but I also think that issue falls little outside the scope of this issue. If the client (in this case Gradle or the plugin) is not authenticated, it is responsibility of the server to not return more information than what client has access to. The client should only show what server has responded with it and not infer meaning and show something different.

**Comment by ljacomet** [comment]
The obvious thing would be to look at the response content. But that could be HTML or anything, which might be hard to then get in an error message in a readable way.
And I am not sure attempting to define a header or some other option for this, from Gradle, makes sense.

If you can show there is already something close to a standard to have more complete errors, please share that.

Without it, this is unlikely to get added to Gradle.

**Comment by assimd** [comment]
Some of the ways we think this could be implemented are:

1. Gradle can define some message schema for error responses, similar to what NPM does (e.g. https://registry.npmjs.org/thisDoesNotExists) and if the response matches the schema, Gradle can deserialization the response body and display the message from server. This might be easiest for Gradle to implement but is also least flexible for repository solution providers to support. If we go this route, we can discuss various options for the schema and format (JSON, XML, etc) ([JFrog's error schema](https://www.jfrog.com/confluence/display/JFROG/Artifactory+REST+API#ArtifactoryRESTAPI-ERRORRESPONSES) might be a good starting point). The schema matching can be lenient, so Gradle will use the response as long as it has the required fields, and ignore everything else. In case of NPM, only expectation is the presence of `error` field, rest everything is swallowed by the client.
2. Gradle can expose the response body/inputStream (optionally with headers) as part of `UnresolvedDependencyResult`, so the plugins can extract the message and display it to the user by logging it. This relaxes the schema restriction on repositories, but now they have to build a plugin for their schema. One potential problem with this approach is that, logging from `afterResolve` hook appears before the `* What went wrong:` section in Gradle output and may get ignored by the users.
3. Gradle can allow plugins to register an error transform callback for each instance of repository. On error, Gradle can invoke this callback (if one is registered) and use the returned value to generate user facing error. e.g `MavenArtifactRepository` can expose a method `setErrorHandler(RepositoryErrorHandler errorHandler)` with definitions as:
    ```
    public interface RepositoryErrorHandler {
        /**
        * Invoked when upstream repository responds with 4XX or 5XX status code.
        * Return `null` to fallback to Gradle's default error handler.
        */
        RepositoryError onRepositoryError(RepositoryResponse repo);
    }

    public class RepositoryResponse {
        private final int httpStatusCode;
        private final Map<String, String> responseHeaders;
        private final InputStream responseBody;
    }

    public RepositoryError {
        private final String errorMessage;
    }
    ```
4. A combination of 1 & 3, where Gradle has a DefaultRepositoryErrorHandler which supports a basic schema, but allow repository solution providers to supply their own `RepositoryErrorHandler` to handle any custom deserialization.

My preference is towards option 4, as it starts with a simple implementation and allow us to add more flexibility later on.

> If you can show there is already something close to a standard to have more complete errors, please share that.

Unfortunately, no two maven repository solution providers share a common error schema, I guess that is because Maven repository protocol never had one defined. IMO, if one such schema is defined by the Gradle, repository providers may either converge on it or if Gradle has mechanism to handle custom schema, they can provide plugins to handle their custom format.

**Comment by assimd** [comment]
@ljacomet, do you think if we can get any of the above design get implemented in Gradle? If so, what additional information do you need before we can start the work? If not, do you have any other approach in mind which can help us solve our problem?

**Comment by vlsi** [comment]
In other words, RFC2616 https://www.w3.org/Protocols/rfc2616/rfc2616-sec6.html seems to be quite standard way of providing server-side message in addition to the status code.

> The first line of a Response message is the Status-Line, consisting of the protocol version followed by a numeric status code and its associated textual phrase, with each element separated by SP characters. No CR or LF is allowed except in the final CRLF sequence.
>       `Status-Line = HTTP-Version SP Status-Code SP Reason-Phrase CRLF`

---

Back to the original example:

> Received status code 403 from server: Forbidden

It means the server provided `Forbidden` as the contents of `Reason-Phrase`.
Then the question boils down to whether you can convince repository developers to provide richer messages than a mere `Forbidden`. Gradle would pick and display it automatically.

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before it gets implemented.

---

A new API would definitely not be implemented. 
As illustrated by Vladimir, we could add a reason phrase as a part of the output, e.g. for cases when it's more than just `Not Found`.

**Comment by lkasso** [comment]
This issue will be closed as a duplicate of

* #36203

Please add your use case and 👍 to that issue.

If you think our analysis is wrong, please provide us with more detailed information explaining why.

***

The newer issue links to a recent RFC and includes a PR in progress.
